src/Internal/Service/Error.purs

```diff
@@ -0,0 +1,97 @@
+module Ctl.Internal.Service.Error
+  ( BlockfrostError(BlockfrostError)
+  , ClientError
+      ( ClientHttpError
+      , ClientHttpResponseError
+      , ClientDecodeJsonError
+      , ClientEncodingError
+      , ClientOtherError
+      )
+  , ServiceError
+      ( ServiceBlockfrostError
+      , ServiceOtherError
+      )
+  ) where
+
+import Prelude
+
+import Aeson (class DecodeAeson, JsonDecodeError, getField)
+import Affjax (Error, printError) as Affjax
+import Affjax.StatusCode (StatusCode) as Affjax
+import Ctl.Internal.Service.Helpers (aesonObject)
+import Data.Generic.Rep (class Generic)
+import Data.Newtype (class Newtype)
+import Data.Show.Generic (genericShow)
+
+--------------------------------------------------------------------------------
+-- ClientError
+--------------------------------------------------------------------------------
+
+data ClientError
+  -- | Affjax error
+  = ClientHttpError Affjax.Error
+  -- | Server responded with HTTP status code outside of 200-299
+  | ClientHttpResponseError Affjax.StatusCode ServiceError
+  -- | Failed to decode the response
+  | ClientDecodeJsonError String JsonDecodeError
+  -- | Failed to encode the request
+  | ClientEncodingError String
+  -- | Any other error
+  | ClientOtherError String
+
+-- No `Show` instance of `Affjax.Error`
+instance Show ClientError where
+  show (ClientHttpError err) =
+    "(ClientHttpError "
+      <> Affjax.printError err
+      <> ")"
+  show (ClientHttpResponseError statusCode err) =
+    "(ClientHttpResponseError "
+      <> show statusCode
+      <> " "
+      <> show err
+      <> ")"
+  show (ClientDecodeJsonError jsonStr err) =
+    "(ClientDecodeJsonError (" <> show jsonStr <> ") "
+      <> show err
+      <> ")"
+  show (ClientEncodingError err) =
+    "(ClientEncodingError "
+      <> err
+      <> ")"
+  show (ClientOtherError err) =
+    "(ClientOtherError "
+      <> err
+      <> ")"
+
+--------------------------------------------------------------------------------
+-- ServiceError
+--------------------------------------------------------------------------------
+
+data ServiceError
```

Comment:
I have a couple of thoughts about this.

First - it looks like all `ServiceOterError` is used by `handleAffjaxResponse` in Kupo. So maybe it should be `KupoError`? Or add `service` field to it, to track which service has failed.
Second - do we plan to actually handle such errors or just crash? If we handle them - then `ClientError` should probably be parametrized by type of `ServiceError` instead of using ADT cases, cuz only one case seems possible, and using cases complicates handling errors.

Reply:
>  parametrized by type of `ServiceError`

I thought about doing that too, but ideally users write contracts which do not care about the choice of backend, which would then lead to type variables everywhere to abstract over the backend... probably not worth it